File: backend/app/api/v1/incidents.py
"""Incident management endpoints."""

import json
import logging
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, desc, or_

from app.core.database import get_db_session
from app.models.incident import Incident, Dispatch
from app.models.vehicle import Vehicle
from app.models.medical import MedicalProfile, EmergencyContact
from app.models.spatial import Hospital, PoliceStation
from app.schemas.incident import TriggerIncidentRequest, AbortIncidentRequest, IncidentOut
from app.services.ai_severity import compute_severity
from app.services.geo_service import (
    get_nearest_hospitals, get_nearest_police_stations, get_driving_route_and_eta
)
from app.services.sms_service import dispatch_emergency_sms
from app.services.websocket_manager import websocket_manager

log = logging.getLogger(__name__)
router = APIRouter(prefix="/incidents")


async def run_emergency_pipeline(
    db: AsyncSession,
    device_id: str,
    vehicle_id: int | None,
    severity_level: str,
    reason: str,
    peak_g_force: str,
    speed_at_impact: str,
    lat: float,
    lng: float,
    ai_score: int = 0,
    ai_summary: str = "",
) -> str:
    """Core emergency response pipeline. Returns incident_ref."""
    import time
    incident_ref = f"INC-{int(time.time() * 1000) % 1_000_000:06d}"

    # Fetch vehicle + medical profile
    vehicle = None
    if vehicle_id:
        result = await db.execute(select(Vehicle).where(Vehicle.id == vehicle_id))
        vehicle = result.scalar_one_or_none()

    medical = None
    if vehicle_id:
        result = await db.execute(
            select(MedicalProfile).where(MedicalProfile.user_id == vehicle.owner_id)
        )
        medical = result.scalar_one_or_none()

    # Patient snapshot for ER teams
    patient_snapshot = {}
    if vehicle:
        patient_snapshot = {
            "name": vehicle.driver_name or "Unknown",
            "bloodGroup": vehicle.blood_group or "Unknown",
            "vehiclePlate": vehicle.registration_number,
        }
    if medical:
        patient_snapshot.update({
            "name": medical.full_name or patient_snapshot.get("name"),
            "bloodGroup": medical.blood_group or patient_snapshot.get("bloodGroup"),
            "allergies": medical.allergies or "None",
            "conditions": medical.medical_conditions or "None",
            "physicianPhone": medical.primary_physician_phone,
            "abhaId": medical.abha_id,
        })

    # Nearest hospital + police
    hospital_list = await get_nearest_hospitals(db, lat, lng, limit=4)
    police_list = await get_nearest_police_stations(db, lat, lng, limit=4)

    nearest_hosp = hospital_list[0] if hospital_list else None
    nearest_pol = police_list[0] if police_list else None

    # Road ETA via OSRM
    hosp_route = {"driving_distance_km": 4.2, "eta_minutes": 8, "polyline": None}
    pol_route = {"driving_distance_km": 2.1, "eta_minutes": 4, "polyline": None}
    if nearest_hosp:
        _, h = nearest_hosp
        hosp_route = await get_driving_route_and_eta(lat, lng, h.lat, h.lng)
    if nearest_pol:
        _, p = nearest_pol
        pol_route = await get_driving_route_and_eta(lat, lng, p.lat, p.lng)

    # AI severity if not provided
    if not ai_score:
        sev = compute_severity(total_g=4.0)
        ai_score = sev.score
        ai_summary = sev.triage_summary

    # Create incident record
    location_name = f"GPS: {lat:.4f}\u00b0N, {lng:.4f}\u00b0E"
    if nearest_hosp:
        _, h = nearest_hosp
        location_name = f"{round(nearest_hosp[0], 1)} km from {h.name}"

    incident = Incident(
        incident_ref=incident_ref,
        vehicle_id=vehicle_id,
        device_id=device_id,
        severity=severity_level,
        reason=reason,
        peak_g_force=peak_g_force,
        speed_at_impact=speed_at_impact,
        location_name=location_name,
        lat=lat,
        lng=lng,
        status="Emergency Active",
        ai_severity_score=ai_score,
        ai_summary=ai_summary,
        patient_snapshot=json.dumps(patient_snapshot),
    )
    db.add(incident)
    await db.flush()

    # Create dispatch record
    hosp_name = None
    hosp_lat = None
    hosp_lng = None
    if nearest_hosp:
        _, h = nearest_hosp
        hosp_name = h.name
        hosp_lat = h.lat
        hosp_lng = h.lng

    pol_name = None
    pol_lat = None
    pol_lng = None
    if nearest_pol:
        _, p = nearest_pol
        pol_name = p.name
        pol_lat = p.lat
        pol_lng = p.lng

    fir_number = f"FIR-{datetime.now(timezone.utc).year}-ASAAS-{int(incident.id):04d}"
    dispatch_logs = [
        f"[{datetime.now(timezone.utc).isoformat()}] High-G Impact event from {device_id}.",
        f"[{datetime.now(timezone.utc).isoformat()}] Nearest Hospital: {hosp_name} "
        f"({hosp_route['driving_distance_km']} km, ETA {hosp_route['eta_minutes']} mins).",
        f"[{datetime.now(timezone.utc).isoformat()}] Nearest Police: {pol_name} "
        f"({pol_route['driving_distance_km']} km, ETA {pol_route['eta_minutes']} mins).",
    ]

    dispatch = Dispatch(
        incident_id=incident.id,
        hospital_name=hosp_name,
        hospital_lat=hosp_lat,
        hospital_lng=hosp_lng,
        ambulance_status="dispatched",
        ambulance_unit="ALS 108 - Trauma Mobile ICU",
        ambulance_eta_minutes=hosp_route["eta_minutes"],
        icu_bed_number="Trauma Bay #04",
        icu_bed_reserved=True,
        blood_units_reserved=2,
        blood_type=patient_snapshot.get("bloodGroup", "Unknown"),
        police_station_name=pol_name,
        police_lat=pol_lat,
        police_lng=pol_lng,
        pcr_status="dispatched",
        pcr_unit="Highway Patrol PCR",
        pcr_eta_minutes=pol_route["eta_minutes"],
        green_corridor_active=True,
        hazard_perimeter_set=True,
        fir_generated=True,
        fir_number=fir_number,
        dispatch_logs=json.dumps(dispatch_logs),
    )
    db.add(dispatch)
    await db.flush()

    # WebSocket broadcast
    broadcast_payload = {
        "incident": {
            "id": incident.id,
            "incidentRef": incident_ref,
            "severity": severity_level,
            "reason": reason,
            "peakGForce": peak_g_force,
            "speedAtImpact": speed_at_impact,
            "location": location_name,
            "coordinates": {"lat": lat, "lng": lng},
            "status": "Emergency Active",
            "aiSeverityScore": ai_score,
            "aiSummary": ai_summary,
            "patientSnapshot": patient_snapshot,
        },
        "dispatches": {
            "hospital": {
                "hospitalName": hosp_name,
                "ambulanceStatus": "dispatched",
                "ambulanceUnit": "ALS 108 - Trauma Mobile ICU",
                "ambulanceEtaMinutes": hosp_route["eta_minutes"],
                "icuBedReserved": True,
                "icuBedNumber": "Trauma Bay #04",
                "bloodUnitsReserved": 2,
                "bloodType": patient_snapshot.get("bloodGroup"),
            },
            "police": {
                "policeStationName": pol_name,
                "pcrStatus": "dispatched",
                "pcrUnit": "Highway Patrol PCR",
                "pcrEtaMinutes": pol_route["eta_minutes"],
                "greenCorridorActive": True,
                "firGenerated": True,
                "firNumber": fir_number,
            },
        },
    }
    await websocket_manager.broadcast("INCIDENT_TRIGGERED", broadcast_payload)

    # SMS (async, non-blocking)
    import asyncio
    if medical:
        result = await db.execute(
            select(EmergencyContact).where(
                EmergencyContact.profile_id == medical.id,
                EmergencyContact.notify_sms == True,
            )
        )
        contacts = result.scalars().all()
        asyncio.create_task(
            dispatch_emergency_sms(
                recipients=[{"phone": c.phone} for c in contacts],
                vehicle_plate=patient_snapshot.get("vehiclePlate", "Unknown"),
                coordinates={"lat": lat, "lng": lng},
                severity=severity_level,
                blood_group=patient_snapshot.get("bloodGroup", "Unknown"),
                nearest_hospital=hosp_name,
            )
        )

    return incident_ref


@router.post("/trigger")
async def trigger_incident(
    body: TriggerIncidentRequest,
    db: AsyncSession = Depends(get_db_session),
):
    sev = compute_severity(total_g=4.0, sos_pressed=(body.severity == "CRITICAL"))
    incident_ref = await run_emergency_pipeline(
        db=db,
        device_id=body.device_id,
        vehicle_id=None,
        severity_level=body.severity,
        reason=body.reason,
        peak_g_force=body.peak_g_force,
        speed_at_impact=body.speed_at_impact,
        lat=body.coordinates.lat,
        lng=body.coordinates.lng,
        ai_score=sev.score,
        ai_summary=sev.triage_summary,
    )
    return {"status": "SUCCESS", "incident_ref": incident_ref}


@router.post("/abort")
async def abort_incident(
    body: AbortIncidentRequest,
    db: AsyncSession = Depends(get_db_session),
):
    if body.incident_ref:
        result = await db.execute(
            select(Incident).where(Incident.incident_ref == body.incident_ref)
        )
        incident = result.scalar_one_or_none()
        if incident:
            incident.status = "ABORTED_FALSE_ALARM"
            incident.resolved_at = datetime.now(timezone.utc)
    else:
        result = await db.execute(
            select(Incident).where(Incident.status == "Emergency Active").order_by(desc(Incident.id))
        )
        incidents = result.scalars().all()
        for i in incidents:
            i.status = "ABORTED_FALSE_ALARM"
            i.resolved_at = datetime.now(timezone.utc)

    await websocket_manager.broadcast(
        "INCIDENT_ABORTED",
        {"reason": body.reason, "timestamp": datetime.now(timezone.utc).isoformat()},
    )
    return {"status": "SUCCESS", "message": "Emergency alert cancelled"}


@router.get("/active")
async def get_active_incident(db: AsyncSession = Depends(get_db_session)):
    result = await db.execute(
        select(Incident)
        .where(Incident.status == "Emergency Active")
        .order_by(desc(Incident.id))
        .limit(1)
    )
    incident = result.scalar_one_or_none()
    if not incident:
        return {"status": "SUCCESS", "incident": None}

    dispatch_result = await db.execute(
        select(Dispatch).where(Dispatch.incident_id == incident.id)
    )
    dispatch = dispatch_result.scalar_one_or_none()

    return {
        "status": "SUCCESS",
        "incident": _format_incident(incident, dispatch),
    }


@router.get("/history")
async def get_incident_history(db: AsyncSession = Depends(get_db_session), limit: int = 20):
    result = await db.execute(
        select(Incident).order_by(desc(Incident.id)).limit(limit)
    )
    incidents = result.scalars().all()
    return {"status": "SUCCESS", "count": len(incidents), "data": [_format_incident(i) for i in incidents]}


@router.get("/{incident_ref}/cap")
async def get_cap_alert(incident_ref: str, db: AsyncSession = Depends(get_db_session)):
    """Export incident as CAP v1.2 (Common Alerting Protocol) JSON."""
    result = await db.execute(
        select(Incident).where(Incident.incident_ref == incident_ref)
    )
    incident = result.scalar_one_or_none()
    if not incident:
        raise HTTPException(status_code=404, detail="Incident not found")

    patient = incident.patient_data
    cap = {
        "identifier": f"IN-ASAAS-{incident.incident_ref}",
        "sender": "emergency-cad@asaas.gov.in",
        "sent": incident.created_at.isoformat(),
        "status": "Actual",
        "msgType": "Alert",
        "scope": "Public",
        "code": ["IPAWS-1.0", "ERSS-112-CAD"],
        "info": [{
            "category": "Safety",
            "event": "Severe Motor Vehicle Collision (MVC)",
            "urgency": "Immediate",
            "severity": "Extreme" if incident.severity == "CRITICAL" else "Severe",
            "certainty": "Observed",
            "headline": f"Automated Impact Detected: {patient.get('vehiclePlate', 'Unknown')} ({incident.peak_g_force})",
            "description": incident.ai_summary or "Sensor collision detected above 4.0G threshold.",
            "instruction": (
                f"Trauma resuscitation required. "
                f"Blood Type: {patient.get('bloodGroup', 'Unknown')}. "
                f"Allergies: {patient.get('allergies', 'None')}." 
            ),
            "area": [{"areaDesc": incident.location_name or "Unknown", "circle": [f"{incident.lat},{incident.lng},0.05"]}],
            "parameter": [
                {"valueName": "VehiclePlate", "value": patient.get("vehiclePlate", "Unknown")},
                {"valueName": "SpeedAtImpact", "value": incident.speed_at_impact or "Unknown"},
                {"valueName": "DecelerationPeak", "value": incident.peak_g_force or "Unknown"},
                {"valueName": "AISeverityScore", "value": str(incident.ai_severity_score)},
            ],
        }],
    }
    return cap


def _format_incident(incident: Incident, dispatch: Dispatch = None) -> dict:
    d = {
        "id": incident.id,
        "incidentRef": incident.incident_ref,
        "vehicleId": incident.vehicle_id,
        "deviceId": incident.device_id,
        "severity": incident.severity,
        "reason": incident.reason,
        "peakGForce": incident.peak_g_force,
        "speedAtImpact": incident.speed_at_impact,
        "locationName": incident.location_name,
        "coordinates": {"lat": incident.lat, "lng": incident.lng},
        "status": incident.status,
        "aiSeverityScore": incident.ai_severity_score,
        "aiSummary": incident.ai_summary,
        "patientSnapshot": incident.patient_data,
        "createdAt": incident.created_at.isoformat() if incident.created_at else None,
        "resolvedAt": incident.resolved_at.isoformat() if incident.resolved_at else None,
        "dispatch": None,
    }
    if dispatch:
        logs = []
        if dispatch.dispatch_logs:
            try:
                logs = json.loads(dispatch.dispatch_logs)
            except Exception:
                logs = []
        d["dispatch"] = {
            "hospitalName": dispatch.hospital_name,
            "ambulanceStatus": dispatch.ambulance_status,
            "ambulanceUnit": dispatch.ambulance_unit,
            "ambulanceEtaMinutes": dispatch.ambulance_eta_minutes,
            "icuBedNumber": dispatch.icu_bed_number,
            "icuBedReserved": dispatch.icu_bed_reserved,
            "bloodUnitsReserved": dispatch.blood_units_reserved,
            "bloodType": dispatch.blood_type,
            "policeStationName": dispatch.police_station_name,
            "pcrStatus": dispatch.pcr_status,
            "pcrUnit": dispatch.pcr_unit,
            "pcrEtaMinutes": dispatch.pcr_eta_minutes,
            "greenCorridorActive": dispatch.green_corridor_active,
            "firGenerated": dispatch.fir_generated,
            "firNumber": dispatch.fir_number,
            "logs": logs,
        }
    return d

File: server/src/routes/incidents.js
import express from 'express';
import { getDb } from '../config/db.js';
import { calculateHaversineDistanceKm, getDrivingRouteAndEta } from '../services/geoSpatialService.js';
import { dispatchEmergencySms } from '../services/smsService.js';
import { socketService } from '../services/socketService.js';

const router = express.Router();

/**
 * Core Emergency Pipeline execution function (Called via REST or Telemetry Edge Crash)
 */
export async function triggerEmergencyPipeline({ deviceId, severity, reason, peakGForce, speedAtImpact, coordinates }) {
  const db = getDb();
  const incidentId = `INC-${Date.now().toString().slice(-6)}`;
  const timestamp = new Date().toISOString();

  // 1. Fetch matching vehicle and medical profile
  const vehicle = await db.get('SELECT * FROM vehicles WHERE device_id = ? OR id = ?', [deviceId, 'v1']) || {
    id: 'v1',
    name: 'Hyundai Creta SX (O) Turbo',
    registration_number: 'DL-01-AB-4321',
    driver_name: 'Aaradhya Sharma',
    blood_group: 'O+ (Positive)'
  };

  const medicalProfile = await db.get('SELECT * FROM medical_profiles LIMIT 1') || {
    full_name: vehicle.driver_name,
    blood_group: vehicle.blood_group,
    allergies: 'Penicillin',
    medical_conditions: 'Mild Asthmatic',
    primary_physician_phone: '+91 98112 34567'
  };

  // 2. Geospatial Query: Find Nearest Hospital
  const hospitals = await db.all('SELECT * FROM hospitals');
  let nearestHosp = null;
  let minHospDist = Infinity;

  for (const h of hospitals) {
    const dist = calculateHaversineDistanceKm(coordinates.lat, coordinates.lng, h.lat, h.lng);
    if (dist < minHospDist) {
      minHospDist = dist;
      nearestHosp = { ...h, distanceKm: dist };
    }
  }

  // 3. Geospatial Query: Find Nearest Police Station
  const policeStations = await db.all('SELECT * FROM police_stations');
  let nearestPolice = null;
  let minPoliceDist = Infinity;

  for (const p of policeStations) {
    const dist = calculateHaversineDistanceKm(coordinates.lat, coordinates.lng, p.lat, p.lng);
    if (dist < minPoliceDist) {
      minPoliceDist = dist;
      nearestPolice = { ...p, distanceKm: dist };
    }
  }

  // 4. Calculate Road Driving ETA via OSRM
  const hospRouting = nearestHosp
    ? await getDrivingRouteAndEta(coordinates.lat, coordinates.lng, nearestHosp.lat, nearestHosp.lng)
    : { drivingDistanceKm: 4.2, etaMinutes: 8 };

  const policeRouting = nearestPolice
    ? await getDrivingRouteAndEta(coordinates.lat, coordinates.lng, nearestPolice.lat, nearestPolice.lng)
    : { drivingDistanceKm: 2.1, etaMinutes: 4 };

  // 5. Store Incident in Database
  const patientSnapshot = JSON.stringify({
    name: medicalProfile.full_name,
    bloodGroup: medicalProfile.blood_group,
    allergies: medicalProfile.allergies,
    conditions: medicalProfile.medical_conditions,
    physicianPhone: medicalProfile.primary_physician_phone
  });

  const locationName = `NH-48 Expressway, KM 34.2 (Near Hero Honda Chowk)`;
  const aiSummary = `${severity || 'CRITICAL'} Impact Collision: Sensor recorded ${peakGForce || '5.84g'}. Nearest facility: ${nearestHosp?.name || 'AIIMS Trauma'}.`;

  await db.run(`
    INSERT INTO incidents (id, vehicle_id, device_id, severity, reason, peak_g_force, speed_at_impact, location_name, lat, lng, patient_snapshot, ai_summary)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `, [
    incidentId,
    vehicle.id,
    deviceId,
    severity || 'CRITICAL',
    reason || 'Accident Crash Detected',
    peakGForce || '5.84g',
    speedAtImpact || '74 km/h',
    locationName,
    coordinates.lat,
    coordinates.lng,
    patientSnapshot,
    aiSummary
  ]);

  // 6. Create Initial Automated Dispatches
  const dispatchId = `DISP-${incidentId}`;
  const initialLogs = JSON.stringify([
    `[${new Date().toLocaleTimeString()}] High-G Impact event received from in-vehicle IoT node (${deviceId}).`,
    `[${new Date().toLocaleTimeString()}] Nearest Hospital Identified: ${nearestHosp?.name} (${hospRouting.drivingDistanceKm} km, ETA: ${hospRouting.etaMinutes} mins).`,
    `[${new Date().toLocaleTimeString()}] Nearest Police Station Identified: ${nearestPolice?.name} (${policeRouting.drivingDistanceKm} km, ETA: ${policeRouting.etaMinutes} mins).`
  ]);

  await db.run(`
    INSERT INTO dispatches (
      id, incident_id, hospital_name, ambulance_status, ambulance_unit, ambulance_eta_minutes,
      icu_bed_number, icu_bed_reserved, blood_units_reserved, blood_type,
      police_station_name, pcr_status, pcr_unit, pcr_eta_minutes, green_corridor_active,
      hazard_perimeter_set, fir_generated, fir_number, logs
    ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
  `, [
    dispatchId,
    incidentId,
    nearestHosp?.name || 'AIIMS Apex Trauma Centre, New Delhi',
    'dispatched',
    'ALS 108 - Trauma Mobile ICU #12',
    hospRouting.etaMinutes,
    'Trauma Bay #04',
    1,
    2,
    medicalProfile.blood_group,
    nearestPolice?.name || 'Sushant Lok Police Station & PCR Hub',
    'dispatched',
    'Highway Patrol Interceptor #07',
    policeRouting.etaMinutes,
    1,
    1,
    1,
    `FIR-2026-DEL-${Math.floor(1000 + Math.random() * 9000)}`,
    initialLogs
  ]);

  // 7. Dispatch Multi-Channel Emergency SMS
  const emergencyContacts = await db.all('SELECT * FROM emergency_contacts WHERE notify_sms = 1');
  dispatchEmergencySms({
    recipients: emergencyContacts,
    vehiclePlate: vehicle.registration_number,
    coordinates,
    severity,
    bloodGroup: medicalProfile.blood_group,
    nearestHospital: nearestHosp?.name
  }).catch(err => console.error('[SMS] Async error:', err));

  const incidentData = {
    id: incidentId,
    date: timestamp.replace('T', ' ').substring(0, 19),
    vehicleName: vehicle.name,
    registrationNumber: vehicle.registration_number,
    deviceId,
    severity: severity || 'CRITICAL',
    reason,
    peakGForce: peakGForce || '5.84g',
    speedAtImpact: speedAtImpact || '74 km/h',
    location: locationName,
    coordinates,
    status: 'Emergency Active (Dispatches In Progress)',
    ambulanceEta: `${hospRouting.etaMinutes} mins (ALS #12 En Route)`,
    aiSummary,
    patientSnapshot: JSON.parse(patientSnapshot),
    nearestHospital: nearestHosp,
    nearestPolice: nearestPolice,
    routes: {
      hospital: hospRouting,
      police: policeRouting
    }
  };

  // 8. WebSocket Broadcast to Hospital ER, Police CAD, and Guardian Circle
  socketService.broadcast('INCIDENT_TRIGGERED', {
    incident: incidentData,
    dispatches: {
      hospital: {
        ambulanceStatus: 'dispatched',
        ambulanceUnit: 'ALS 108 - Trauma Mobile ICU #12',
        ambulanceEtaMinutes: hospRouting.etaMinutes,
        icuBedReserved: true,
        icuBedNumber: 'Trauma Bay #04',
        bloodUnitsReserved: 2,
        bloodType: medicalProfile.blood_group,
        hospitalName: nearestHosp?.name
      },
      police: {
        pcrStatus: 'dispatched',
        pcrUnit: 'Highway Patrol Interceptor #07',
        pcrEtaMinutes: policeRouting.etaMinutes,
        greenCorridorActive: true,
        hazardPerimeterSet: true,
        firGenerated: true
      }
    }
  });

  return incidentData;
}

/**
 * @route POST /api/v1/incidents/trigger
 * @desc Trigger an emergency manually or from simulated API
 */
router.post('/trigger', async (req, res) => {
  try {
    const { deviceId = 'ASAAS-001', severity = 'CRITICAL', reason = 'Manual SOS Alert', peakGForce = '5.84g', speedAtImpact = '68 km/h', coordinates = { lat: 28.4595, lng: 77.0266 } } = req.body;

    const incident = await triggerEmergencyPipeline({
      deviceId,
      severity,
      reason,
      peakGForce,
      speedAtImpact,
      coordinates
    });

    return res.status(201).json({ status: 'SUCCESS', incident });
  } catch (err) {
    console.error('[INCIDENT] Trigger error:', err);
    return res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

/**
 * @route GET /api/v1/incidents/active
 * @desc Get current active emergency incident
 */
router.get('/active', async (req, res) => {
  try {
    const db = getDb();
    const incident = await db.get("SELECT * FROM incidents WHERE status LIKE '%Emergency Active%' ORDER BY created_at DESC LIMIT 1");
    if (!incident) {
      return res.json({ status: 'SUCCESS', incident: null });
    }

    const dispatch = await db.get('SELECT * FROM dispatches WHERE incident_id = ?', [incident.id]);

    return res.json({
      status: 'SUCCESS',
      incident: {
        ...incident,
        patientSnapshot: incident.patient_snapshot ? JSON.parse(incident.patient_snapshot) : null,
        coordinates: { lat: incident.lat, lng: incident.lng },
        dispatch: dispatch ? {
          ...dispatch,
          logs: dispatch.logs ? JSON.parse(dispatch.logs) : []
        } : null
      }
    });
  } catch (err) {
    return res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

/**
 * @route GET /api/v1/incidents/history
 * @desc Get all past incident records
 */
router.get('/history', async (req, res) => {
  try {
    const db = getDb();
    const incidents = await db.all('SELECT * FROM incidents ORDER BY created_at DESC LIMIT 20');
    return res.json({
      status: 'SUCCESS',
      count: incidents.length,
      data: incidents.map(i => ({
        ...i,
        coordinates: { lat: i.lat, lng: i.lng },
        patientSnapshot: i.patient_snapshot ? JSON.parse(i.patient_snapshot) : null
      }))
    });
  } catch (err) {
    return res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

/**
 * @route POST /api/v1/incidents/abort
 * @desc Abort/Cancel false alarm within countdown
 */
router.post('/abort', async (req, res) => {
  try {
    const { incidentId, reason = 'Driver verified safe via 15s PIN/Button cancellation' } = req.body;
    const db = getDb();

    if (incidentId) {
      await db.run("UPDATE incidents SET status = 'ABORTED_FALSE_ALARM', resolved_at = CURRENT_TIMESTAMP WHERE id = ?", [incidentId]);
    } else {
      await db.run("UPDATE incidents SET status = 'ABORTED_FALSE_ALARM', resolved_at = CURRENT_TIMESTAMP WHERE status LIKE '%Emergency Active%'");
    }

    socketService.broadcast('INCIDENT_ABORTED', { reason, timestamp: new Date().toISOString() });

    return res.json({ status: 'SUCCESS', message: 'Emergency alert canceled successfully' });
  } catch (err) {
    return res.status(500).json({ status: 'ERROR', message: err.message });
  }
});

export default router;
